ntp-proto/src/filter.rs

```diff
@@ -359,28 +357,56 @@ fn construct_candidate_list<'a>(
     candidate_list
 }
 
-/// Find the largest contiguous intersection of correctness
-/// intervals.  Allow is the number of allowed falsetickers;
-/// found is the number of midpoints.  Note that the edge values
-/// are limited to the range +-(2 ^ 30) < +-2e9 by the timestamp
-/// calculations.
 #[allow(dead_code)]
-fn find_interval(chime_list: &[CandidateTuple]) -> (NtpDuration, NtpDuration) {
+struct SurvivorTuple<'a> {
+    p: &'a Peer,
+    metric: NtpDuration,
+}
+
+/// Collect the candidates within the correctness interval
+#[allow(dead_code)]
+fn construct_survivors<'a>(
+    chime_list: &[CandidateTuple<'a>],
+    local_clock_time: NtpTimestamp,
+) -> Vec<SurvivorTuple<'a>> {
+    let mut survivors = Vec::new();
+
+    if let Some((low, high)) = find_interval(chime_list) {
+        for tuple in chime_list {
+            if tuple.edge < low || tuple.edge > high {
+                continue;
+            }
+
+            let p = tuple.peer;
+            let metric = MAX_DISTANCE * p.stratum + p.root_distance(local_clock_time);
+
+            survivors.push(SurvivorTuple { p, metric })
+        }
+    }
+
+    survivors
+}
+
+/// Find the largest contiguous intersection of correctness intervals.
+#[allow(dead_code)]
+fn find_interval(chime_list: &[CandidateTuple]) -> Option<(NtpDuration, NtpDuration)> {
     let n = chime_list.len();
 
-    let mut low = NtpDuration::ONE * 2_000_000_000;
-    let mut high = low * -164;
+    let mut low = None;
+    let mut high = None;
 
+    // allow is the number of allowed falsetickers
     for allow in (0..).take_while(|allow| 2 * allow < n) {
-        // falsetickers found in the current iteration
-        let mut found = 0;
-        let mut chime = 0;
+        let mut found = 0; // variable "d", falsetickers found in the current iteration
+        let mut chime = 0; // variable "c"
 
         // Scan the chime list from lowest to highest to find the lower endpoint.
         for tuple in chime_list {
             chime -= tuple.endpoint_type as i32;
-            if chime >= (n - found) as i32 {
-                low = tuple.edge;
+
+            // the code skeleton uses `n - found` here, which is wrong!
+            if chime >= (n - allow) as i32 {
```

Comment:
this is subtle, but using `found` would just find the same result for every iteration. The specification text uses the `f` variable, which we give the name `allow`. 

Another case where you need to read very carefully.